app/services/service_provider_updater.rb

```diff
@@ -0,0 +1,47 @@
+class ServiceProviderUpdater
+  def run
+    dashboard_service_providers.each do |sp|
+      issuer = sp['issuer'] || sp[:issuer]
+      SERVICE_PROVIDERS[issuer] = sp
+
+      VALID_SERVICE_PROVIDERS << issuer unless VALID_SERVICE_PROVIDERS.include?(issuer)
+
+      # must also alter secure headers
+      # see config/initializers/secure_headers.rb
+      alter_secure_headers(sp)
+    end
+  end
+
+  private
+
+  def url
+    Figaro.env.dashboard_url
+  end
+
+  def dashboard_service_providers
+    body = dashboard_response.body
+    return JSON.parse(body) if dashboard_response.code == 200
+    log_error "Failed to parse response from #{url}: #{body}"
+    []
+  rescue
+    log_error "Failed to contact #{url}"
+    []
+  end
+
+  def dashboard_response
+    @_dashboard_response ||= HTTParty.get(url)
+  end
+
+  def alter_secure_headers(sp)
```

Comment:
The CSP header is not for non-https. It's for whitelisting domain names that our SAML Authn "POST" form automatically submits to when users log in.
Look at `app/views/saml_idp/shared/saml_post_binding.html.slim`